File: frontend/src/services/hotel-service.ts
import apiClient from '@/lib/api';
import { Hotel, HotelCreate, HotelUpdate, HotelResponse } from '@/types/hotel';

export const hotelService = {
  // Get all hotels (public access)
  async getHotels(params?: { skip?: number; limit?: number; active_only?: boolean }): Promise<Hotel[]> {
    const response = await apiClient.get('/hotels/', { params });
    return response.data;
  },

  // Get a specific hotel by ID
  async getHotel(hotelId: string): Promise<Hotel> {
    const response = await apiClient.get(`/hotels/${hotelId}`);
    return response.data;
  },

  // Create a new hotel (admin access required)
  async createHotel(hotelData: HotelCreate): Promise<Hotel> {
    const response = await apiClient.post('/hotels/', hotelData);
    return response.data;
  },

  // Update a hotel (admin access required)
  async updateHotel(hotelId: string, hotelData: HotelUpdate): Promise<Hotel> {
    const response = await apiClient.put(`/hotels/${hotelId}`, hotelData);
    return response.data;
  },

  // Delete a hotel (admin access required)
  async deleteHotel(hotelId: string): Promise<void> {
    await apiClient.delete(`/hotels/${hotelId}`);
  },

  // Search hotels by city or country
  async searchHotels(params?: { city?: string; country?: string }): Promise<Hotel[]> {
    const response = await apiClient.get('/hotels/search/', { params });
    return response.data;
  },

  // Toggle hotel active status
  async toggleHotelActive(hotelId: string, isActive: boolean): Promise<Hotel> {
    const response = await apiClient.put(`/hotels/${hotelId}`, { is_active: isActive });
    return response.data;
  },

  async getMyHotels()  {
    const response = await apiClient.get('/hotels/myHotels');
    return response.data;
  },

  // Get cheapest room price for a hotel
  async getCheapestRoomPrice(hotelId: string): Promise<number | null> {
    try {
      const response = await apiClient.get(`/rooms/hotel/${hotelId}`);
      const rooms = response.data;
      
      if (!Array.isArray(rooms) || rooms.length === 0) {
        return null;
      }
      
      // Find the minimum price from available rooms
      const prices = rooms
        .filter((room: any) => {
          return room && 
                 typeof room.price_per_night === 'number' && 
                 room.price_per_night > 0;
        })
        .map((room: any) => room.price_per_night);
      
      return prices.length > 0 ? Math.min(...prices) : null;
    } catch (error) {
      console.error(`Error fetching cheapest room price for hotel ${hotelId}:`, error);
      return null;
    }
  }
};

File: backend/app/services/hotel_service.py
from typing import List, Optional
from datetime import datetime
from beanie import PydanticObjectId
from app.models.hotel import Hotel, HotelCreate, HotelUpdate, HotelResponse


class HotelService:
    @staticmethod
    async def create_hotel(hotel_data: HotelCreate, creator_id: Optional[str] = None) -> HotelResponse:
        """Create a new hotel"""
        hotel_dict = hotel_data.model_dump()
        if creator_id:
            hotel_dict["created_by"] = creator_id

        hotel = Hotel(**hotel_dict)
        await hotel.create()
        
        # Use model_validate to create response from hotel document
        return HotelResponse.model_validate({
            **hotel.model_dump(),
            "id": str(hotel.id)
        })

    @staticmethod
    async def get_hotel(hotel_id: str) -> Optional[HotelResponse]:
        """Get a hotel by ID"""
        try:
            hotel = await Hotel.get(PydanticObjectId(hotel_id))
            if hotel:
                return HotelResponse.model_validate({
                    **hotel.model_dump(),
                    "id": str(hotel.id)
                })
        except Exception:
            return None
        return None

    @staticmethod
    async def get_hotels(skip: int = 0, limit: int = 100, active_only: bool = True) -> List[HotelResponse]:
        """Get all hotels with pagination"""
        query = Hotel.find()
        if active_only:
            query = Hotel.find(Hotel.is_active == True)
        
        hotels = await query.skip(skip).limit(limit).to_list()
        
        return [
            HotelResponse.model_validate({
                **hotel.model_dump(),
                "id": str(hotel.id)
            })
            for hotel in hotels
        ]

    @staticmethod
    async def update_hotel(hotel_id: str, hotel_data: HotelUpdate) -> Optional[HotelResponse]:
        """Update a hotel"""
        try:
            hotel = await Hotel.get(PydanticObjectId(hotel_id))
            if not hotel:
                return None

            update_data = {k: v for k, v in hotel_data.model_dump(exclude_unset=True).items() if v is not None}
            
            if update_data:
                await hotel.update({"$set": update_data})
                
                # Fetch updated hotel
                updated_hotel = await Hotel.get(PydanticObjectId(hotel_id))
                return HotelResponse.model_validate({
                    **updated_hotel.model_dump(),
                    "id": str(updated_hotel.id)
                })
        except Exception:
            return None
        return None

    @staticmethod
    async def delete_hotel(hotel_id: str) -> bool:
        """Delete a hotel (soft delete by setting is_active to False)"""
        try:
            hotel = await Hotel.get(PydanticObjectId(hotel_id))
            if hotel:
                await hotel.update({"$set": {"is_active": False}})
                return True
        except Exception:
            pass
        return False

    @staticmethod
    async def search_hotels(city: Optional[str] = None, country: Optional[str] = None) -> List[HotelResponse]:
        """Search hotels by city or country"""
        query = {"is_active": True}
        
        if city:
            query["city"] = {"$regex": city, "$options": "i"}
        if country:
            query["country"] = {"$regex": country, "$options": "i"}
        
        hotels = await Hotel.find(query).to_list()
        
        return [
            HotelResponse.model_validate({
                **hotel.model_dump(),
                "id": str(hotel.id)
            })
            for hotel in hotels
        ]
    

    @staticmethod
    async def get_hotels_by_creator(creator_id: str) -> List[HotelResponse]:
            hotels = await Hotel.find(Hotel.created_by == creator_id).to_list()
            
            return [
                HotelResponse.model_validate({
                    **hotel.model_dump(),
                    "id": str(hotel.id)
                })
                for hotel in hotels
            ]
